### PR TITLE
Add api scopes

### LINKED ISSUE #69: Add support for creating 'Okta API scopes'
*@rachrdfds commented on Dec 2, 2019, 3:01 PM UTC:*

**Is your feature request related to a problem? Please describe.**
When creating an application that is to be used to create and maintain other Okta applications etc. in Okta, some Okta API scopes must be enabled/granted on the application.

Currently I can create the application using Terraform but the I ahve to set the Okta API scopes manually (GUI):
[image](https://user-images.githubusercontent.com/57186882/69969333-856cf780-151c-11ea-8782-30cf63d2aa62.png)

**Describe the solution you'd like**
A way to create the application with the specified Okta API scopes. It could be an array of strings containing one or more of (below is currently a complete list):

*   okta.apps.manage
*   okta.apps.read
*   okta.authorizationServers.manage
*   okta.authorizationServers.read
*   okta.clients.manage
*   okta.clients.read
*   okta.clients.register
*   okta.eventHooks.manage
*   okta.eventHooks.read
*   okta.events.read
*   okta.factors.manage
*   okta.factors.read
*   okta.groups.manage
*   okta.groups.read
*   okta.idps.manage
*   okta.idps.read
*   okta.inlineHooks.manage
*   okta.inlineHooks.read
*   okta.logs.read
*   okta.roles.manage
*   okta.roles.read
*   okta.schemas.manage
*   okta.schemas.read
*   okta.users.manage
*   okta.users.manage.self
*   okta.users.read
*   okta.users.read.self

**Describe alternatives you've considered**
I don't see any workaround for this.

*This issue was moved by [noinarisak](https://github.com/noinarisak) from [articulate/terraform-provider-okta#361](https://github.com/articulate/terraform-provider-okta/issues/361).*

### PR DESCRIPTION
**Pull request by mariussturm**
This PR adds a new resource `okta_app_oauth_api_scope` in order to manage OAuth API scopes.
First-time provider dev here might be that there is some oddity.

Fixes #69, #237

### CODE REVIEW

**Comment by bogdanprodan-okta** [comment]
Hi, @mariussturm! Thanks for submitting this PR. I'll review it and get back to asap.

**Comment by mcrobbj** [comment]
Is this going to be released?

**Comment by bogdanprodan-okta** [comment]
> Is this going to be released?

Hi! Yes, this is going to be released by the end of this week.

File: okta/resource_okta_app_oauth_api_scope.go [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+			"issuer": {
+				Required:    true,
+				Type:        schema.TypeString,
+				Description: "The issuer of your Org Authorization Server, your Org URL.",
+			},
+			"scopes": {
+				Type:        schema.TypeList,
+				Required:    true,
+				Elem:        &schema.Schema{Type: schema.TypeString},
+				Description: "Scopes of the application for which consent is granted.",
```
Comment (bogdanprodan-okta) [inline okta/resource_okta_app_oauth_api_scope.go:81]: Could you please add validation for the scopes based on the list from [here](https://developer.okta.com/docs/guides/implement-oauth-for-okta/scopes/)?

File: okta/resource_okta_app_oauth_api_scope.go [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+	for _, scope := range d.Get("scopes").([]interface{}) {
+		desiredScopes = append(desiredScopes, scope.(string))
+	}
+
+	// extract scope list form []okta.OAuth2ScopeConsentGrant
+	for _, currentScope := range from {
+		currentScopes = append(currentScopes, currentScope.ScopeId)
+	}
+
+	// find scopes that should not be there
```
Comment (bogdanprodan-okta) [inline okta/resource_okta_app_oauth_api_scope.go]: As an advice, you can use `splitTargets` (okta/resource_okta_admin_role_targets.go:183) func here. Yes, the naming is confusing, so feel free to rename params.

**Review by bogdanprodan-okta — CHANGES_REQUESTED** [review]

File: okta/resource_okta_app_oauth_api_scope.go (hunk 1 above)
Comment (mariussturm) [reply okta/resource_okta_app_oauth_api_scope.go:81]: Totally agree, validation would be good here but it looks like tf doesn't like validation on lists. I get the following error:

```
1 error occurred:
	* resource okta_app_oauth_api_scope: scopes: ValidateFunc and
ValidateDiagFunc are not yet supported on lists or sets.
```

I could put the validation in the setter function but that becomes a bit messy, do you have any idea?

File: okta/resource_okta_app_oauth_api_scope.go (hunk 2 above)
Comment (mariussturm) [reply okta/resource_okta_app_oauth_api_scope.go]: Ah good catch, we have the same code already. I Will change that!

File: okta/resource_okta_app_oauth_api_scope.go (hunk 1 above)
Comment (bogdanprodan-okta) [reply okta/resource_okta_app_oauth_api_scope.go:81]: Try putting validation for each element `Elem:        &schema.Schema{Type: schema.TypeString},`

Comment (mariussturm) [reply okta/resource_okta_app_oauth_api_scope.go:81]: Ha, stupid me, yes that works perfectly. Thanks again!

**Comment by bogdanprodan-okta** [comment]
The last thing I wanted to ask is to merge with the latest master and run `make fmt`. I've introduced a new more advanced [formatting tool](https://github.com/mvdan/gofumpt) instead of standard `gofmt`.

**Review by bogdanprodan-okta — APPROVED** [review]

**Comment by mariussturm** [comment]
Thanks for the approval! Did a rebase and `make fmt`.

**Comment by blakemann** [comment]
@bogdanprodan-okta I've got a project that I'm waiting to be able to use this in - now that it's merged, will there be another release being done soon to include it?

**Comment by bogdanprodan-okta** [comment]
Hi, @blakemann ! The next release will probably this week. The only problem is that provider has migrated from `oktadeveloper` to `okta` and I don't know if I will be able to release it with the current `gpg` key. 
There is always a possibility to use [in-house provider ](https://www.terraform.io/docs/cloud/run/install-software.html#in-house-providers) built from source. Basically, you run `go install`, copy the binary into your `terraform.d` dir and run `terraform init`.

**Comment by mariussturm** [comment]
@bogdanprodan-okta thanks for the investigation! A registry deployment of 3.11 would be great so that we can officially make use of that feature.

**Comment by tmatilai** [comment]
Sorry to comment to an old PR, but is there any reason why the `issuer` attribute couldn't be internally fetched from the provider configuration?

On the client side, the provider attributes are not accessible (especially problem on common modules), and there doesn't seem to exist any data source for getting the URL either.

/cc @bogdanprodan-okta

**Comment by mariussturm** [comment]
@tmatilai if the `issuer` would be fetched from the internal configuration it has to be the org URL under all circumstances and all edge cases. I am not sure if this is always true, especially when custom authorization servers come into play.

What exactly is the use-case for reading back the `issuer` URL? Usually, you know the Okta base URL in the tf code and just render it here again?

**Comment by tmatilai** [comment]
> if the `issuer` would be fetched from the internal configuration it has to be the org URL under all circumstances and all edge cases.

If there are edge cases, then the `issuer` attribute could be optional, and default to the provider configuration. 🤷 

> What exactly is the use-case for reading back the `issuer` URL? Usually, you know the Okta base URL in the tf code and just render it here again?

If the provider is passed from a root module to a common module, the common module doesn't know about the org or base URL any more. So now it's required to add separate variables just for the `issue` URL. 

In case of auth server is created, the base URL can actually be parsed from its issuer URL, but that feels extraneous, too.

Btw, thanks @mariussturm for adding the resource in the first place! 🤗

**Pull request merged**